app/controllers/ApiController.scala

```diff
@@ -258,4 +261,64 @@ final class ApiController @Inject()(api: OreRestfulApi,
     */
   def showStatusZ = Action(Ok(this.status.json))
 
+  def syncSso() = Action { implicit request =>
+    this.forms.SyncSso.bindFromRequest.fold(
+      hasErrors => BadRequest(Json.obj("errors" -> hasErrors.errorsAsJson)),
+      formData => {
+        val sso = formData._1
+        val sig = formData._2
+        val apiKey = formData._3
+
+
+        val confApiKey = this.config.security.get[String]("sync_sso.apikey")
+        val confSecret = this.config.security.get[String]("sync_sso.secret")
+
+        if (apiKey != confApiKey) {
+          BadRequest("API Key not valid")
+        } else if (!CryptoUtils.hmac_sha256(confSecret, sso.getBytes("UTF-8")).equals(sig)) {
+          BadRequest("Signature not matched")
+        } else {
+          val decoded = URLDecoder.decode(new String(Base64.getMimeDecoder.decode(sso)), "UTF-8")
```

Comment:
Play (through Akka) has nicer ways of dealing with query strings. Without looking too much into it, something like `Uri.Query(Base64.getMimeDecoder.decode(sso))` (charset is UTF-8 by default) is what you want to go with. Then call toMap on the result of that. From there, no more vars. Also as another tip, Scala provides a `toInt` method on strings to parse them as ints.